version/bump_from_params_test.go

```diff
@@ -27,10 +28,11 @@ var _ = Describe("BumpForParams", func() {
 
 		bumpParam = ""
 		preParam = ""
+		buildParam = ""
```

Comment:
:thought_balloon: Looks like this var has been added but I don't see any contexts that set it to anything. So the tests below just seem to test the behavior of applying a `bump:` to a version that already has build info, but there's nothing showing how build info is added. Seems like that test could be added pretty easily?

Reply:
I've added more tests.  Please review and let me know if you see more that should be added.

Reply:
I may be missing something but I still don't see a context here that shows what happens when you set `build:` in the params to *add* build data to the version. :thinking: Am I misinterpreting the PR? So far the tests just assert that any existing build data on the version is preserved.

I'm actually kinda confused as to what the user-facing goal of this PR is - what's here makes sense but I thought the original ask was to support adding metadata to versions as part of bumping. Right now it's supported but only via static configuration, which may be somewhat limiting as it means the build data is determined at pipeline configuration time, rather than at runtime (so that you can generate it based on things like git commits).

Could you post a mock-up of a pipeline you'd like to define demonstrating this functionality, just so we're on the same page?

Thanks! (And thanks for tackling this!)

Reply:
Honestly, I just tried to follow the same pattern that I saw the other tests use.  This was my first attempt at GoLang, so I very well might have not understood a lot of how the versioning happens.  I am also new to Concourse and certainly not as familiar with how it works.

What I was attempting to accomplish is to get build number and git hash appended onto the version.  Our strategy on Jenkins pipeline is to generate a version number like this:

1.0.0+b1.1a2b3c4d 

I assumed ( probably incorrectly ) that there was a build variable available to the pipeline and that I could somehow generate the git hash of the code repository into a variable or file, and then pass those to the version resource like so:

jobs:
- name: bump-version
  serial: true
  plan:
  - get: version
    params: {bump: patch, build: "b"+BUILD_NUM+"."+GIT_HASH} //not sure of exact syntax
  - put: version
    params: {file: version/version}

Am I even thinking about this correctly?

Thanks!

Reply:
Ah, nope that won't do what you want. Concourse doesn't expose any magic variables for that, in fact it doesn't even really know what it's doing anything `git`-specific so it wouldn't really be able to.

I think for this feature you'd need to bump the version via `put`, because generating the version will depend on information obtained at runtime. `get` steps can't reference runtime information because they're effectively fully resolved by the the time the build runs (in fact they're usually what *determines* whether the build will run, via `trigger: true`, so this is kind of fundamental.)

So you'd want something like:

```yaml
- get: my-repo
- put: version
  params:
    bump: patch
    build_file: my-repo/.git/short_ref
    build_prefix: "ref."
```

This would generate a version like `1.2.3+ref.ac1beef`, using the `.git/short_ref` file created by the [`git` resource](https://github.com/concourse/git-resource#additional-files-populated).

If you want the build number in there, the resource would have to explicitly support interpolating the env vars. For this you could use Go's [`os.Expand`](https://golang.org/pkg/os/#ExpandEnv) function on the string value. Then you could do this:

```yaml
- get: my-repo
- put: version
  params:
    bump: patch
    build_file: my-repo/.git/short_ref
    build_prefix: "b${BUILD_NAME}."
```

This would work because Concourse exposes a few special [metadata env vars](https://concourse-ci.org/implementing-resources.html#resource-metadata) available only to resources (not tasks). A version generated may then look like `1.2.3+b456.ac1beef` if it were run by the 456'th build of a job.

So for this to work, your PR would need to add support for `build_file` and `build_prefix` params to `/out`. The `build_file` param would work by loading the contents of the specified file (be careful to strip whitespace) and adding it as the version's build metadata, with the optional `build_prefix` param prepended to it (after running it through `os.Expand`).

Does that make sense? Happy to answer any more questions if you need help.